6X pg_dump: Backport performance improvements

bmdoil opened the pull request:
This PR is backporting a subset of the PG upstream performance improvements to pg_dump from https://github.com/greenplum-db/gpdb/pull/13791

Relevant changes from upstream are detailed in the individual commit messages.

soumyadeep2007 commented on src/bin/pg_dump/pg_dump_sort.c:
```diff
@@ -31,46 +31,88 @@ static const char *modulename = gettext_noop("sorter");
-	21,							/* DO_BLOB */
-	24,							/* DO_BLOB_DATA */
-	8,							/* DO_EXTPROTOCOL */
-	21,							/* DO_TYPE_STORAGE_OPTIONS */
```
`DO_TYPE_STORAGE_OPTIONS` was left over from `6cae730feb1`, so its exclusion is not a problem from the final list.

soumyadeep2007 commented on src/bin/pg_dump/pg_backup.h:
```diff
@@ -69,6 +69,37 @@ typedef enum _teSection
+/* We need one enum entry per prepared query in pg_dump */
+enum _dumpPreparedQueries
+{
+	PREPQUERY_DUMPAGG,
+	PREPQUERY_DUMPBASETYPE,
+	PREPQUERY_DUMPCOMPOSITETYPE,
+	PREPQUERY_DUMPDOMAIN,
+	PREPQUERY_DUMPENUMTYPE,
+	PREPQUERY_DUMPFUNC,
+	PREPQUERY_DUMPOPR,
+	PREPQUERY_DUMPRANGETYPE,
+	PREPQUERY_DUMPTABLEATTACH,
+	PREPQUERY_GETCOLUMNACLS,
+	PREPQUERY_GETDOMAINCONSTRAINTS,
```
Hmm, seems like we didn't carry over changes to some of the object types. Eg: `PREPQUERY_DUMPAGG`, `PREPQUERY_DUMPBASETYPE`, `PREPQUERY_DUMPRANGETYPE` apart from `PREPQUERY_DUMPTABLEATTACH`

Is there a specific reason or is this by accident?

bmdoil replied:
Mistake, fixed.

soumyadeep2007 commented:
I reviewed all of the commits except the following:

```
pg_dump: Reorganize getTables()
Backported from 47bfbd4

Conflicts resolved:
* WIP
...
```

Can you fill in the WIP? And then I will take another look.

bmdoil commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -1362,18 +1255,12 @@ setup_connection(Archive *AH, const char *dumpencoding, char *use_role)
 		ExecuteSqlStatement(AH,
 							"SET TRANSACTION ISOLATION LEVEL "
-							"SERIALIZABLE READ ONLY");
+							"SERIALIZABLE, READ ONLY");
```
`* note: comma was not accepted in SET TRANSACTION before 8.0 */` 
Since our min version is now 8.3 this can always read `SERIALIZABLE, READ ONLY`

soumyadeep2007 replied:
Ah gotcha.

bmdoil commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -9739,47 +9649,50 @@ dumpFunc(Archive *fout, FuncInfo *finfo)
+							"proiswin as proiswindow,\n");
+
+	/* GPDB6 added proexeclocation */
+	if (fout->remoteVersion >= 90400)
```
Makes sense, done.

bmdoil commented:
> I reviewed all of the commits except the following:
> 
> ```
> pg_dump: Reorganize getTables()
> Backported from 47bfbd4
> 
> Conflicts resolved:
> * WIP
> ...
> ```
> 
> Can you fill in the WIP? And then I will take another look.
Updated the commit message and addressed remaining feedback.

soumyadeep2007 reviewed (approved):
LGTM!

kyeap-vmware commented:
So we don't lose the context. We'll want to revert commit.
https://github.com/greenplum-db/gpdb/pull/14164/commits/2793dae44fb8e0a15de567047403566a459e6282
